crates/noirc_abi/src/serialization.rs

```diff
@@ -0,0 +1,233 @@
+    }
+}
+
+impl AbiType {
+    fn from_type(param_type: Type, visibility: AbiFEType) -> Self {
+        match param_type {
+            Type::Field => AbiType::Field(visibility),
+            Type::Integer { sign, width } => AbiType::Integer { visibility, sign, width },
+            Type::Array { length, typ } => AbiType::Array {
+                visibility,
+                length: length.into(),
+                typ: Box::new(Self::from_type(*typ, visibility)),
+            },
+            Type::Struct { fields } => AbiType::Struct {
+                visibility,
+                fields: fields
+                    .into_iter()
+                    .map(|field| (field.name, AbiType::from_type(field.param_type, visibility)))
+                    .collect(),
+            },
+        }
+    }
+}
+
+impl Serialize for Abi {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let parameter_vec: Vec<AbiParameter> = self
+            .parameters
+            .clone()
+            .into_iter()
+            .map(|(name, param_type)| {
+                let visibility = Some(param_type.visibility());
+                AbiParameter { name, visibility, param_type: Type::from(param_type) }
+            })
+            .collect();
+
+        parameter_vec.serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for Abi {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let parameters = Vec::<AbiParameter>::deserialize(deserializer)?;
+
+        let parameters = parameters
+            .into_iter()
+            .map(|AbiParameter { name, visibility, param_type }| {
+                (name, AbiType::from_type(param_type, visibility.unwrap()))
+            })
+            .collect();
+
+        Ok(Abi { parameters })
+    }
+
+    fn deserialize_in_place<D>(deserializer: D, place: &mut Self) -> Result<(), D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        // Default implementation just delegates to `deserialize` impl.
+        *place = Deserialize::deserialize(deserializer)?;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::BTreeMap;
+
+    use crate::{Abi, AbiFEType, AbiType, Sign};
+
+    #[test]
+    fn abi_deserializes_correctly() {
+        let serialized_abi = "
+        [
+            {
+                \"name\": \"thing1\",
+                \"visibility\": \"public\",
+                \"type\": {
+                    \"kind\": \"field\"
+                }
+            },
+            {
+                \"name\": \"thing2\",
+                \"visibility\": \"private\",
+                \"type\": {
+                    \"kind\": \"array\",
+                    \"length\": 2,
+                    \"type\": {
+                        \"kind\": \"integer\",
+                        \"width\": 3,
+                        \"sign\": \"unsigned\"
```

Comment:
I wonder if it'd be at all possible to accept `u3` here to better reflect what users would actually write in noir.
Especially with compound types like `[u3; 3]` however we'd need a separate parser for these values after the json string is read. So perhaps it is impractical if this is only meant to be read/written to automatically by programs anyway.

Reply:
Yeah, I suggested this alternative in https://github.com/noir-lang/noir/issues/524. I feel that if we include structs in the ABI then this could be pretty messy unless we had a separate definitions section, which is also yucky.

Reply:
> So perhaps it is impractical if this is only meant to be read/written to automatically by programs anyway.

Yes, that's right. I'm expecting this file to be read to either generate circuit-specific interfaces for handling inputs for different circuits (The [Typechain for noir](https://github.com/dethcrypto/TypeChain) idea), or reload the `Abi` struct without recomputing from source.

Reply:
On second thoughts, we don't need to describe the structure of the struct in the `kind` field as structs are self-describing,  we could then just leave them as `"kind": "struct"`.

It still leaves the issue that we'd need to perform the extra parsing step on each `kind` field though.